kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -499,348 +534,173 @@ protected void handleInactive(KafkaHeaderAndRequest kafkaHeaderAndRequest,
+                new ArrayList<>(adminManager.getBrokers(advertisedEndPoint.getListenerName())));
 
-        // 2. After get all topics, for each topic, get the service Broker for it, and add to response
-        AtomicInteger topicsCompleted = new AtomicInteger(0);
         // Each Pulsar broker can manage metadata like controller in Kafka,
         // Kafka's AdminClient needs to find a controller node for metadata management.
         // So here we return an random broker as a controller for the given listenerName.
         final int controllerId = adminManager.getControllerId(advertisedEndPoint.getListenerName());
-        pulsarTopicsFuture.whenComplete((pulsarTopics, e) -> {
+
+        final String namespacePrefix = currentNamespacePrefix();
+        final MetadataRequest request = (MetadataRequest) metadataHar.getRequest();
+        // This map is used to find the original topic name. Both key and value don't have the "-partition-" suffix.
+        final Map<String, String> fullTopicNameToOriginal = (request.topics() == null)
+                ? Collections.emptyMap()
+                : request.topics().stream().collect(
+                        Collectors.toMap(
+                                topic -> new KopTopic(topic, namespacePrefix).getFullName(),
+                                topic -> topic
+                        ));
+        // NOTE: for all topics METADATA request, remove the default namespace prefix just for backward compatibility.
+        final Function<String, String> getOriginalTopic = fullTopicName -> fullTopicNameToOriginal.isEmpty()
+                ? KopTopic.removeDefaultNamespacePrefix(fullTopicName, namespacePrefix)
+                : fullTopicNameToOriginal.getOrDefault(fullTopicName, fullTopicName);
+
+        final String metadataNamespace = kafkaConfig.getKafkaMetadataNamespace();
+        getTopicsAsync(request, fullTopicNameToOriginal.keySet()).whenComplete((topicAndMetadataList, e) -> {
             if (e != null) {
-                log.warn("[{}] Request {}: Exception fetching metadata, will return null Response",
-                    ctx.channel(), metadataHar.getHeader(), e);
-                MetadataResponse finalResponse =
-                        KafkaResponseUtils.newMetadata(
-                                allNodes,
-                                clusterName,
-                                controllerId,
-                                Collections.emptyList());
-                resultFuture.complete(finalResponse);
+                log.error("[{}] Request {}: Exception fetching metadata, will return null Response",
+                        ctx.channel(), metadataHar.getHeader(), e);
+                resultFuture.complete(KafkaResponseUtils.newMetadata(
+                        allNodes,
+                        clusterName,
+                        controllerId,
+                        Collections.emptyList()));
                 return;
             }
```

Comment:
One of my concerns is that if handle permission validation with an unrecoverable error, instead of having no permissions, the client won't be able to receive the exception from the server, even though it doesn't care about the exception, but since the client just got an empty TopicMetadata , the client will continue to retry the MetadataRequest , and a large number of MetadataRequest will consume more CPU of the broker. In addition, the client may not find that there is a problem with its own program.

I looked at the implementation of handleMetadataRequest in kafka. If an broker level fault is encountered, the broker will return the error to the client.

Reply:
It changes the current behavior as well. I think if we can handle this exceptional case in another PR so that we can add some tests to verify it.

Reply:
After rethinking for a while, I think we need to complete it exceptionally here. See the latest commit.

Reply:
Good work! Thanks. :)